File: pipelines/basalt/examples/m7bw_temporal_probe.rs
//! M7bw external temporal KLT trace for MH_01 frame 0 -> frame 1, cam0 track 2.
//!
//! This probe is intentionally outside the production stream. It exposes the
//! public patch/update primitives at every level/iteration so their f32 bits
//! can be compared with the pinned Eigen/Sophus probe.

use std::{env, path::PathBuf};

use nalgebra::{Matrix2, Vector2};
use visloc_basalt::{
    AffineCompact2f, DirectKltConfig, EurocSensorDataset, MeanNormalizedPatch51, RawU16Pyramid, Se2,
};

fn bits(value: f32) -> String {
    format!("0x{:08x}", value.to_bits())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut args = env::args_os().skip(1);
    let root = PathBuf::from(args.next().ok_or("dataset root")?);
    let calib = PathBuf::from(args.next().ok_or("calibration")?);
    let config_path = PathBuf::from(args.next().ok_or("config")?);
    let dataset = EurocSensorDataset::open(root, calib, config_path)?;
    let frame0 = dataset.frame(0)?;
    let frame1 = dataset.frame(1)?;
    let old_pyramid = RawU16Pyramid::from_image(frame0.cam0, 3)?;
    let current_pyramid = RawU16Pyramid::from_image(frame1.cam0, 3)?;
    let config = DirectKltConfig {
        pyramid_levels: 3,
        max_iterations: 5,
        ..DirectKltConfig::default()
    };
    let source = Vector2::new(46.0_f32, 118.0_f32);
    let mut transform = AffineCompact2f::new(Matrix2::identity(), source);

    println!("source={},{}", bits(source.x), bits(source.y));
    for level in (0..=config.pyramid_levels).rev() {
        let scale = (1_u32 << level) as f32;
        let old_position = source / scale;
        let old_image = old_pyramid.level(level).expect("old level");
        let current_image = current_pyramid.level(level).expect("current level");
        let patch = MeanNormalizedPatch51::from_image(old_image, old_position);
        println!(
            "L{level}.mean={} data0={} jac0={},{},{} valid={}",
            bits(patch.mean),
            bits(patch.data[0]),
            bits(patch.jacobian_se2[(0, 0)]),
            bits(patch.jacobian_se2[(0, 1)]),
            bits(patch.jacobian_se2[(0, 2)]),
            patch.valid
        );
        for row in 0..3 {
            print!("L{level}.inv{row}=");
            for sample in 0..52 {
                if sample != 0 {
                    print!(",");
                }
                print!("{}", bits(patch.h_se2_inv_j_se2_t[(row, sample)]));
            }
            println!();
        }
        let mut level_transform =
            AffineCompact2f::new(*transform.linear(), *transform.translation() / scale);
        for iteration in 0..config.max_iterations {
            let residual = patch
                .residual(current_image, &level_transform)
                .expect("temporal residual");
            print!("L{level}.I{iteration}.res=");
            for (index, value) in residual.iter().enumerate() {
                if index != 0 {
                    print!(",");
                }
                print!("{}", bits(*value));
            }
            println!();
            let increment = patch.ic_increment(&residual);
            println!(
                "L{level}.I{iteration}.inc={},{},{}",
                bits(increment.x),
                bits(increment.y),
                bits(increment.z)
            );
            level_transform.right_compose_se2(increment);
            println!(
                "L{level}.I{iteration}.t={},{}",
                bits(level_transform.translation().x),
                bits(level_transform.translation().y)
            );
            println!(
                "L{level}.I{iteration}.m={},{},{},{},{},{}",
                bits(level_transform.linear()[(0, 0)]),
                bits(level_transform.linear()[(0, 1)]),
                bits(level_transform.linear()[(1, 0)]),
                bits(level_transform.linear()[(1, 1)]),
                bits(level_transform.translation().x),
                bits(level_transform.translation().y)
            );
            let update = Se2::exp(increment);
            println!(
                "L{level}.I{iteration}.upd={},{},{},{},{},{}",
                bits(update.rotation[(0, 0)]),
                bits(update.rotation[(0, 1)]),
                bits(update.rotation[(1, 0)]),
                bits(update.rotation[(1, 1)]),
                bits(update.translation.x),
                bits(update.translation.y)
            );
            if !current_image.in_bounds(*level_transform.translation(), 2.0) {
                break;
            }
        }
        transform = AffineCompact2f::new(
            *level_transform.linear(),
            *level_transform.translation() * scale,
        );
        println!(
            "L{level}.world_t={},{}",
            bits(transform.translation().x),
            bits(transform.translation().y)
        );
        println!(
            "L{level}.world_m={},{},{},{},{},{}",
            bits(transform.linear()[(0, 0)]),
            bits(transform.linear()[(0, 1)]),
            bits(transform.linear()[(1, 0)]),
            bits(transform.linear()[(1, 1)]),
            bits(transform.translation().x),
            bits(transform.translation().y)
        );
    }
    println!(
        "final={},{}",
        bits(transform.translation().x),
        bits(transform.translation().y)
    );
    Ok(())
}

File: benchmarks/basalt/m7bw_temporal_probe.cpp
// M7bw external native trace for MH_01 frame 0 -> frame 1, cam0 track 2.
// This duplicates only Basalt's public FrameToFrameOpticalFlow kernel so the
// pinned Eigen/Sophus intermediates can be compared with the Rust probe.

#include <basalt/image/image_pyr.h>
#include <basalt/optical_flow/patch.h>
#include <basalt/optical_flow/patterns.h>
#include <basalt/utils/vio_config.h>

#include <opencv2/imgcodecs.hpp>

#include <cstdint>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>

using Affine = Eigen::AffineCompact2f;
using Patch = basalt::OpticalFlowPatch<float, basalt::Pattern51<float>>;

static std::uint32_t bits(float value) {
  std::uint32_t result = 0;
  std::memcpy(&result, &value, sizeof(result));
  return result;
}

static void print_bits(float value) {
  std::cout << "0x" << std::hex << std::setw(8) << std::setfill('0')
            << bits(value) << std::dec << std::setfill(' ');
}

static basalt::ManagedImage<uint16_t> read_image(const std::string& path) {
  cv::Mat image = cv::imread(path, cv::IMREAD_UNCHANGED);
  if (image.empty() || image.type() != CV_8UC1) {
    throw std::runtime_error("expected 8-bit grayscale image: " + path);
  }
  basalt::ManagedImage<uint16_t> result(image.cols, image.rows);
  for (int y = 0; y < image.rows; ++y) {
    const auto* source = image.ptr<std::uint8_t>(y);
    auto* destination = result.RowPtr(y);
    for (int x = 0; x < image.cols; ++x) {
      destination[x] = static_cast<std::uint16_t>(source[x]) << 8;
    }
  }
  return result;
}

static void print_vector(const Patch::VectorP& vector) {
  for (int index = 0; index < Patch::PATTERN_SIZE; ++index) {
    if (index != 0) std::cout << ',';
    print_bits(vector[index]);
  }
  std::cout << '\n';
}

int main(int argc, char** argv) {
  if (argc < 3) return 2;
  const std::string root = argv[1];
  const std::string config_path = argv[2];
  basalt::VioConfig config;
  config.load(config_path);

  const std::int64_t timestamps[] = {1403636579763555584LL,
                                     1403636579813555456LL};
  std::ostringstream old_name;
  old_name << root << "/mav0/cam0/data/" << timestamps[0] << ".png";
  std::ostringstream current_name;
  current_name << root << "/mav0/cam0/data/" << timestamps[1] << ".png";
  basalt::ManagedImage<uint16_t> old_image = read_image(old_name.str());
  basalt::ManagedImage<uint16_t> current_image = read_image(current_name.str());
  basalt::ManagedImagePyr<uint16_t> old_pyramid, current_pyramid;
  old_pyramid.setFromImage(old_image, config.optical_flow_levels);
  current_pyramid.setFromImage(current_image, config.optical_flow_levels);

  const Eigen::Vector2f source(46.0f, 118.0f);
  Affine transform;
  transform.setIdentity();
  transform.translation() = source;
  std::cout << "source=";
  print_bits(source.x());
  std::cout << ',';
  print_bits(source.y());
  std::cout << '\n';

  for (int level = config.optical_flow_levels; level >= 0; --level) {
    const float scale = static_cast<float>(1 << level);
    transform.translation() /= scale;
    Patch patch(old_pyramid.lvl(level), source / scale);
    std::cout << "L" << level << ".mean=";
    print_bits(patch.mean);
    std::cout << " data0=";
    print_bits(patch.data[0]);
    std::cout << " jac0=";
    print_bits(patch.H_se2_inv_J_se2_T(0, 0));
    std::cout << " valid=" << patch.valid << '\n';
    for (int row = 0; row < 3; ++row) {
      std::cout << "L" << level << ".inv" << row << "=";
      for (int sample = 0; sample < Patch::PATTERN_SIZE; ++sample) {
        if (sample != 0) std::cout << ',';
        print_bits(patch.H_se2_inv_J_se2_T(row, sample));
      }
      std::cout << '\n';
    }

    for (int iteration = 0;
         iteration < config.optical_flow_max_iterations && patch.valid;
         ++iteration) {
      Patch::Matrix2P transformed =
          transform.linear().matrix() * Patch::pattern2;
      transformed.colwise() += transform.translation();
      Patch::VectorP residual;
      if (!patch.residual(current_pyramid.lvl(level), transformed, residual)) {
        throw std::runtime_error("native residual failed");
      }
      std::cout << "L" << level << ".I" << iteration << ".res=";
      print_vector(residual);
      std::cout << '\n';
      const Eigen::Vector3f increment = -patch.H_se2_inv_J_se2_T * residual;
      std::cout << "L" << level << ".I" << iteration << ".inc=";
      print_bits(increment[0]);
      std::cout << ',';
      print_bits(increment[1]);
      std::cout << ',';
      print_bits(increment[2]);
      std::cout << '\n';
      const Eigen::Matrix3f update = Sophus::SE2f::exp(increment).matrix();
      transform *= update;
      std::cout << "L" << level << ".I" << iteration << ".t=";
      print_bits(transform.translation().x());
      std::cout << ',';
      print_bits(transform.translation().y());
      std::cout << '\n';
    }
    transform.translation() *= scale;
    std::cout << "L" << level << ".world_t=";
    print_bits(transform.translation().x());
    std::cout << ',';
    print_bits(transform.translation().y());
    std::cout << '\n';
  }
  std::cout << "final=";
  print_bits(transform.translation().x());
  std::cout << ',';
  print_bits(transform.translation().y());
  std::cout << '\n';
  return 0;
}
